### PR TITLE
Stringchanges

### PR DESCRIPTION
**Pull request by gmcode**

### CODE REVIEW

**Comment by gmcode** [comment]
Will try to squash a bit more and resubmit.

**Comment by gmcode** [comment]
Didn't manage to squash so will keep it the same.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.inl [hunk 1]
```diff
@@ -149,6 +150,12 @@ const T* StringBase< T >::CString() const
 }
 
 template< typename T >
+bool StringBase<T>::_IsUsingLocalBuffer() const
```
Comment (DolceTriade) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.inl]: libRocket doesn't use the underscore in front of the function anywhere else. Just leave it as Is

File: src/libs/libRocket/Include/Rocket/Core/StringBase.h [hunk 2]
```diff
@@ -185,8 +189,18 @@ class StringBase
 	inline const T& operator[](size_type index) const;
 	inline T& operator[](size_type index);
 
+	// Offer c_str() to make the interface somewhat more consistent with std::string.
+	// It eases code migration to std::string in the future should that be needed.
+	// The name also imparts better expectations about what it does better
+	// than CString() which can be confused with things like MFC's class of the same name.
+	inline const T* c_str() const;
```
Comment (DolceTriade) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.h]: If its inline, define it inside the StringBase.inl file. inl stands for inline so all the functions in that file should be automatically inlined.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.h [hunk 3]
```diff
@@ -185,8 +189,18 @@ class StringBase
...
+
 protected:	
+	inline T* _LocalBuffer();
```
Comment (DolceTriade) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.h]: For this and the function below, libRocket doesn't use this naming convention for protected/private methods. You should remove the underscore.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.h (hunk 2 above)
Comment (DolceTriade) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.h]: Not necessary as CString() exists.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.inl [hunk 4]
```diff
@@ -165,12 +172,18 @@ void StringBase< T >::Reserve(size_type size)
 	if (value == (T*)local_buffer)
 	{
 		T* new_value = (T*)realloc(NULL, buffer_size);
+		// TODO! Consider throwing here.
```
Comment (DolceTriade) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.inl]: librocket doesn't use exceptions.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.h (hunk 3 above)
Comment (gmcode) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.h]: That's not what I am observing, what about _Assign, _Insert, etc. etc. ?
I'm happy to not use _ but that's why I added it in the first place.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.h (hunk 2 above)
Comment (gmcode) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.h]: If I have to comply I will, but unless there is more to this I disagree. Using CString in the first place was probably a bad idea IMO as It doesn't convey what c_str() and doesn't help generic code and it doesn't serve a useful purpose to propagate the situation. Plus it's longer and defeats muscle memory for regular string. It's is also an inline function that has no cost to add now so why wait.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.inl (hunk 4 above)
Comment (gmcode) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.inl]: I'll discuss this more on irc and comply based on the outcome of that. But the current argument  makes no sense to me.

Is there anywhere that it says librocket shouldn't use exceptions for any reason? If there is none, perhaps we should add some and the justification for it?

But given the rest of stl does use exceptions it seems exceptions will happen somewhere anyway?

Also given there are only two places to add the bad_alloc support to address the issue properly, it seems deficient not to address the issue unless there is good reason. I mean why force any user of this library to crash with no message or a misleading access violation one if there is no good reason, when we can instead with trivial effort leave it up to the app to decide to crash or not by simply catching the exception or not and without a misleading message. That's what they are for, right?

To do otherwise without good reason seems short sited and illogical unless a real "customer" of the library explicitly needs exceptions disabled and even then it should probably a a compile time enabled thing.

In summary, I think the argument for not using exceptions is insufficiently justified so far to agree with.

File: src/libs/libRocket/Include/Rocket/Core/StringBase.h (hunk 3 above)
Comment (DolceTriade) [inline src/libs/libRocket/Include/Rocket/Core/StringBase.h]: It would be better. They seem to only use it when the caller has the same function name.

**Comment by Viech** [comment]
@gmcode it's possible to just update an existing pull request with new or changed commits, no need to close and reopen. Just push your branch again, if I remember correctly. :wink:
